src/dmd/traits.d

```diff
@@ -412,6 +413,31 @@ extern (C++) Expression pointerBitmap(TraitsExp e)
     return ale;
 }
 
+/**
+* __traits(getCallerSource, symbol) => get symbol source code (string[T.length] or string)
+*/
+extern (C++) Expression traitsGetSource(TraitsExp e)
+{
+    if (!e.args || e.args.dim < 1|| e.args.dim > 2)
+    {
+        error(e.loc, "TODO");
+        return new ErrorExp();
+    }
+    auto ret = new ArgStringInitExp(e.loc);
+    auto ident_expr=(*e.args)[0];
+    auto t=isType(ident_expr);
+    if(!t){
+        error(e.loc, "TODO.2");
+        return new ErrorExp();
+    }
+    auto name=ident_expr.toChars();
+    // TODO: lookup?
+    auto id = Identifier.idPool(name, name.strlen);
+    ret.setIdent(id);
+    ret.exp=e;
```

Comment:
Perhaps ArgStringInitExp should have a constructor that takes these two expressions as arguments.